cardano-api/src/Cardano/Api/SerialiseLedgerCddl.hs

```diff
@@ -0,0 +1,318 @@
+      Right $ ShelleyBootstrapWitness era w
+    "Key Witness ShelleyEra" -> do
+      w <- first TextEnvelopeCddlErrCBORDecodingError
+             $ CBOR.decodeAnnotator"Shelley Witness" fromCBOR (LBS.fromStrict teCddlRawCBOR)
+      Right $ ShelleyKeyWitness era w
+    _ -> Left TextEnvelopeCddlUnknownKeyWitness
+
+writeTxFileTextEnvelopeCddl
+  :: IsCardanoEra era
+  => FilePath
+  -> Tx era
+  -> IO (Either (FileError ()) ())
+writeTxFileTextEnvelopeCddl path tx =
+  runExceptT $ do
+    handleIOExceptT (FileIOError path) $ LBS.writeFile path txJson
+ where
+  txJson = encodePretty' textEnvelopeCddlJSONConfig (serialiseTxLedgerCddl tx) <> "\n"
+
+writeTxWitnessFileTextEnvelopeCddl
+  :: ShelleyBasedEra era
+  -> FilePath
+  -> KeyWitness era
+  -> IO (Either (FileError ()) ())
+writeTxWitnessFileTextEnvelopeCddl sbe path w =
+  runExceptT $ do
+    handleIOExceptT (FileIOError path) $ LBS.writeFile path txJson
+ where
+  txJson = encodePretty' textEnvelopeCddlJSONConfig (serialiseWitnessLedgerCddl sbe w) <> "\n"
+
+textEnvelopeCddlJSONConfig :: Config
+textEnvelopeCddlJSONConfig =
+  defConfig { confCompare = textEnvelopeCddlJSONKeyOrder }
+
+textEnvelopeCddlJSONKeyOrder :: Text -> Text -> Ordering
+textEnvelopeCddlJSONKeyOrder = keyOrder ["type", "description", "cborHex"]
+
+-- | This GADT allows us to deserialise a tx or key witness without
+--having to provide the era.
+data FromSomeTypeCDDL c b where
+  FromCDDLTx
+    :: Text -- ^ CDDL type that we want
+    -> (InAnyCardanoEra Tx -> b)
+    -> FromSomeTypeCDDL TextEnvelopeCddl b
+
+  FromCDDLWitness
+    :: Text -- ^ CDDL type that we want
+    -> (InAnyCardanoEra KeyWitness -> b)
+    -> FromSomeTypeCDDL TextEnvelopeCddl b
+
+deserialiseFromTextEnvelopeCddlAnyOf
+  :: [FromSomeTypeCDDL TextEnvelopeCddl b]
+  -> TextEnvelopeCddl
+  -> Either TextEnvelopeCddlError b
+deserialiseFromTextEnvelopeCddlAnyOf types teCddl =
+    case List.find matching types of
+      Nothing ->
+        Left (TextEnvelopeCddlTypeError expectedTypes actualType)
+
+      Just (FromCDDLTx ttoken f) -> do
+        AnyCardanoEra era <- cddlTypeToEra ttoken
+        f . InAnyCardanoEra era <$> deserialiseTxLedgerCddl era teCddl
+
+      Just (FromCDDLWitness ttoken f) -> do
+         AnyCardanoEra era <- cddlTypeToEra ttoken
+         case cardanoEraStyle era of
+           LegacyByronEra -> Left TextEnvelopeCddlErrByronKeyWitnessUnsupported
+           ShelleyBasedEra sbe ->
+             f . InAnyCardanoEra era <$> deserialiseWitnessLedgerCddl sbe teCddl
+  where
+   actualType :: Text
+   actualType = teCddlType teCddl
+
+   expectedTypes :: [Text]
+   expectedTypes = [ typ | FromCDDLTx typ _f <- types ]
+
+   matching :: FromSomeTypeCDDL TextEnvelopeCddl b -> Bool
+   matching (FromCDDLTx ttoken _f) = actualType == ttoken
+   matching (FromCDDLWitness ttoken _f)  = actualType == ttoken
+
+-- TODO: This is atrocious but the plan is to parse the 'teCddlType'
+-- field in the future.
```

Comment:
Is it sufficient to just write `TODO Parse the 'teCddlType' field in the future`, so is the atrocious bit pointing to something worth noting explicity?

Reply:
I updated the comment though :+1: